### PR TITLE
ENH: More helpful/forgiving io.wavfile errors

### PR DESCRIPTION
**Pull request by endolith**
#### Reference issue
Related to #11076, #10332, #4755

#### What does this implement/fix?
Error messages are more explicit about what's wrong, and extraneous bytes at end of file are ignored instead of raising an error when the data has successfully been read (similar to #10870).

For the file in #4755, it now says:

     ValueError: Unknown wave file format: MPEGLAYER3. Supported formats: PCM, IEEE_FLOAT

If you try to read a PDF (similar to #7677), it says 

    ValueError: File format b'%PDF' not understood. Only 'RIFF' and 'RIFX' supported.

If you try to read an AVI (which is also RIFF format), it now says:

    ValueError: Not a WAV file. RIFF form type is b'AVI '.

A file with _PMX, DISP, LIST, and id3 chunks, ~with an extraneous byte at the end~ (see https://github.com/scipy/scipy/pull/12208), says:

    WavFileWarning: Chunk (non-data) not understood, skipping it.
    WavFileWarning: Incomplete chunk ID: b'\x00', ignoring it.

but still reads the data.

List of all formats: https://github.com/voidqk/basicwav/blob/master/docs/wav.md

### CODE REVIEW

**Comment by larsoner** [comment]
Is it possible to update some tests with something like `with pytest.raises(ValueError, match='...')` to make it clear better messages are being emitted?

**Comment by endolith** [comment]
Yes I can add some tests.  I also added the complete list of formats, which I will probably push to this PR, and then probably an `info()` function that would return the bit depth, format name, chunk types, etc. of a file, which I'll do in a different PR

The Travis failure is bogus?

**Comment by rgommers** [comment]
> The Travis failure is bogus?

indeed, please ignore - something in the new Numba or Sparse releases is unhappy.

**Comment by endolith** [comment]
@larsoner 

> something like with pytest.raises(ValueError, match='...') 

Should they just match the entire error message?

**Comment by endolith** [comment]
Ok, updated tests!

- The .ani file was the smallest RIFF file I could find that isn't WAVE.  Feel free to suggest something else, though, like a 2-frame low-res AVI or something.
- 3COM_NBX isn't a valid variable name, so I used underscore `_3COM_NBX` because Google returns things for that, while `THREECOM_NBX` only finds https://formats.kaitai.io/wav/  Alternatively they could all keep the `WAVE_FORMAT_3COM_NBX` names, but that seems excessive.
- is `del data` inside the tests necessary? If so, why don't they all have it?

File: scipy/io/tests/test_wavfile.py [hunk 1]
```diff
@@ -75,36 +85,76 @@ def test_read_5():
...
+            assert "CDF" in str(excinfo.value)
+            assert "'RIFF' and 'RIFX' supported" in str(excinfo.value)
+
+
+def test_read_unknown_riff_form_type():
+    # RIFF, but not WAVE form
+    for mmap in [False, True]:
+        filename = 'Transparent Busy.ani'
+        with open(datafile(filename), 'rb') as fp:
+            with raises(ValueError) as excinfo:
```
Comment (larsoner) [inline scipy/io/tests/test_wavfile.py:105]: I would just do `raises(ValueError, match='Not a WAV file.*ACON'):` where `match` is a suitable regex match. I usually try to match something that's specific enough to ensure we're catching the right thing but also not too long (no need to do the entire message).

**Review by larsoner — COMMENTED** [review]
Other than a test clarity suggestion, LGTM

**Comment by larsoner** [comment]
> is del data inside the tests necessary? If so, why don't they all have it?

No idea, maybe it was useful for memory purposes at some point?

File: scipy/io/tests/test_wavfile.py (hunk 1 above)
Comment (endolith) [reply scipy/io/tests/test_wavfile.py:105]: Ok, that makes them awkward though, like 

`with raises(ValueError, match='Unknown wave file format.*MULAW.*Supported formats'):`

I wanted to just test for a few components and allow the wording to be changed, etc. without affecting the test, but I don't know if that really matters

**Comment by endolith** [comment]
> No idea, maybe it was useful for memory purposes at some point?

The file's only 1 KB, and wouldn't garbage collection get it at that point anyway?

File: scipy/io/tests/test_wavfile.py (hunk 1 above)
Comment (larsoner) [reply scipy/io/tests/test_wavfile.py:105]: > that makes them awkward though, like

That looks fine to me, quicker to read and more compact than the `excinfo`. I prefer it personally

> I wanted to just test for a few components and allow the wording to be changed, etc. without affecting the test, but I don't know if that really matters

If someone changes the wording such that it no longer matches it's fairly trivial to change the test to accommodate it, so I don't think we need to worry about this too much

Comment (endolith) [reply scipy/io/tests/test_wavfile.py]: Ok I updated them

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge. Let's give a couple of days for others to have a chance to look then merge. Thanks in advance @endolith !

**Comment by larsoner** [comment]
Thanks @endolith

**Comment by tylerjereddy** [comment]
I added a short release note for this since it is listed as an enhancement: 
https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.5.0#scipyio-improvements

Please feel free to adjust/improve/remove as you see fit.

**Comment by tylerjereddy** [comment]
One other thing here--there are at least 3 cross-listed issues that are still open; if you feel they should be closed, please do so along with addition of the appropriate release milestone.

**Pull request merged**